### ISSUE TITLE
Spinner disappears when using toast.promise() and custom icons on ToastContainers via callback

### ISSUE THREAD

**Issue by paulleonartcalvo** [fkhadra/react-toastify#849]
**Do you want to request a *feature* or report a *bug*?**
This is a bug

**What is the current behavior?**
If no callback is provided to `ToastContainer.icon`, then the spinner appears as expected in any toast which is created via `toast.promise()`. However, if a callback is provided to `ToastContainer.icon`, the following combinations are not possible:

1) a custom icon for a specific toaster `type` AND a spinner for that `type` when using promise, _without_ having to override at the caller level
2) no custom icon defined for that `type` within a callback AND a spinner for that `type` when using promise,  _without_ having to override at the caller level

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your CodeSandbox (https://codesandbox.io/s/new) example below:**

1) Set up a ToastContainer
2) Provide a callback for ToasterContainer.icon. Mine looks like the following: 

```
<ToastContainer
        ...
          icon={(props) => {
            const iconScale = theme.iconScale.small;
            const iconColor = props.theme === 'dark' ? theme.colors['neutral-0'] : theme.colors['neutral-75'];
            switch (props.type) {
              case 'info':
                return <InfoOutlineIcon size={iconScale} color={iconColor} />;
              case 'success':
                return <CheckmarkOutlineIcon size={iconScale} color={iconColor} />;
              case 'error':
                return <CautionOutlineIcon size={iconScale} color={iconColor} />;
              case 'warning':
                return <CircleCautionOutlineIcon size={iconScale} color={iconColor} />;
              case 'default':
                return <NotificationOutlineIcon size={iconScale} color={iconColor} />;
              default:
                return undefined;
            }
          }}
...
</ToastContainer>
```
2) Create a toast via `toast.promise`. My call looks like this: 

```
const prom = triggerUploadImage({ body: formData }).unwrap(); // This is Promise<ApiData>
toast.promise(
  prom,
  {
    pending: 'Uploading image',
    success: { type: 'success', render: 'Image published' },
    error: { type: 'error', render: 'Something went wrong while publishing the image' }
  },
  { toastId: 'upload-image-toast' }
);
```

The result is a toast which is of the default type (as expected), but does not show the spinner even though it 1) is called via `toast.promise()` and 2) does not have a custom icon defined (as i would expect undefined to behave in that callback)

**Here is a sandbox showing the behavior**: [Link](https://codesandbox.io/s/mui-react-toastify-spinner-bug-hsi2zu)

**What is the expected behavior?**

When no Icon override or custom style is defined (not `undefined`) for a specific `type` of toast, or if the return of that function is undefined, then no override should take place, and the default icon should show when toast is created **without** `toast.promise` and no other options which would override the icon, if using a custom icon, unless otherwise specified the spinner should still be used when the promise is running. This makes sense because in `toast()`, setting `options.icon` to undefined does not de-render the icon, it simply doesn't provide an override. Either there needs to be more props showing whether the toast was created via `toast.promise`, or there needs to be more explicit options regarding this overriding behavior. 

For example, maybe the callback can return a object similar to the toast options like so: 

```
ToastContainer.icon`: (props: {type: string; theme: string}) => {
renderIcon: ReactNode; // here is where the icon is rendered
iconReplacesSpinner?: boolean = false; // default false, if true then it replaces the spinner, regardless of renderSpinner
renderSpinner?: ReactNode} // undefined meaning the default spinner is used
```
This way i can define a custom icon, and whether it should replace the spinner. And regardless of if replaces the spinner or not, i can define a custom spinner render.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
As far as i know this is not caused by a React version. I am using macOS Monterey 12.5.1 and React 17.0.2

**Comment by ZainGulbaz** [comment]
Hi @fkhadra, Is the issue still open?

**Comment by ZainGulbaz** [comment]
@paulleonartcalvo You have explained the issue and its possible solution really well. I appreciate that and if the issue is still open I can start working on it.

**Comment by paulleonartcalvo** [comment]
This issue appears to still be open :/

**Comment by fkhadra** [comment]
Hey this issue will be addressed by https://github.com/fkhadra/react-toastify/issues/1003 without introducing a new props. Instead in the function you can do as follow

```
import {Icons} from "react-toastify"

<ToastContainer
        ...
          icon={(props) => {
            const iconScale = theme.iconScale.small;
            const iconColor = props.theme === 'dark' ? theme.colors['neutral-0'] : theme.colors['neutral-75'];

            if (props.isLoading) return <Icons.spinner/>

            switch (props.type) {
              case 'info':
                return <InfoOutlineIcon size={iconScale} color={iconColor} />;
              case 'success':
                return <CheckmarkOutlineIcon size={iconScale} color={iconColor} />;
              case 'error':
                return <CautionOutlineIcon size={iconScale} color={iconColor} />;
              case 'warning':
                return <CircleCautionOutlineIcon size={iconScale} color={iconColor} />;
              case 'default':
                return <NotificationOutlineIcon size={iconScale} color={iconColor} />;
              default:
                return undefined;
            }
          }}
...
</ToastContainer>

```

**Comment by augustosnk12** [comment]
> Hey this issue will be addressed by [#1003](https://github.com/fkhadra/react-toastify/issues/1003) without introducing a new props. Instead in the function you can do as follow
> 
> ```
> import {Icons} from "react-toastify"
> 
> <ToastContainer
>         ...
>           icon={(props) => {
>             const iconScale = theme.iconScale.small;
>             const iconColor = props.theme === 'dark' ? theme.colors['neutral-0'] : theme.colors['neutral-75'];
> 
>             if (props.isLoading) return <Icons.spinner/>
> 
>             switch (props.type) {
>               case 'info':
>                 return <InfoOutlineIcon size={iconScale} color={iconColor} />;
>               case 'success':
>                 return <CheckmarkOutlineIcon size={iconScale} color={iconColor} />;
>               case 'error':
>                 return <CautionOutlineIcon size={iconScale} color={iconColor} />;
>               case 'warning':
>                 return <CircleCautionOutlineIcon size={iconScale} color={iconColor} />;
>               case 'default':
>                 return <NotificationOutlineIcon size={iconScale} color={iconColor} />;
>               default:
>                 return undefined;
>             }
>           }}
> ...
> </ToastContainer>
> ```

This worked like a charm! Thanks @fkhadra  🥰
